File: Week_2/P2.cpp
class FooBar {
private:
    int n;
    mutex gMutex;
    condition_variable cv;
    int turn;

public:
    FooBar(int n) {
        this->n = n;
        this->turn = 0;
    }

    void foo(function<void()> printFoo) {
        for (int i = 0; i < n; i++) {
            {
                unique_lock lk(gMutex);
                cv.wait(lk, [this](){return turn == 0;});
            }
        	printFoo();
            {
                unique_lock lk(gMutex);
                turn = 1;
            }
            cv.notify_all();
        }
    }

    void bar(function<void()> printBar) {
        for (int i = 0; i < n; i++) {
            {
                unique_lock lk(gMutex);
                cv.wait(lk, [this](){return turn == 1;});
            }
        	printBar();
            {
                unique_lock lk(gMutex);
                turn = 0;
            }
            cv.notify_all();
        }
    }
};

File: Week_2/P2.java
class FooBar {
    private int n;
    private int turn;

    public FooBar(int n) {
        this.n = n;
        this.turn = 0;
    }

    public void foo(Runnable printFoo) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            synchronized(this) {
                while(turn != 0) {
                    wait();
                }
            }
            printFoo.run();
            synchronized(this) {
                turn = 1;
                notifyAll();
            }
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        for (int i = 0; i < n; i++) {
            synchronized(this) {
                while(turn != 1) {
                    wait();
                }
            }
        	printBar.run();
            synchronized(this) {
                turn = 0;
                notifyAll();
            }
        }
    }
}